### PR TITLE
fix(dataset-validator): read metadata via read_elem, not backed read_h5ad (#447)

### PR DESCRIPTION
**Pull request by NoopDog**
Closes #447.

## Problem

`read_file_metadata` opened the whole h5ad via `anndata.read_h5ad(path, backed="r")` just to read `obs` + `uns`. anndata's backed mode is **X-only** — `read_h5ad_backed` (anndata `_io/h5ad.py`) eagerly `read_elem`s `layers` and `raw`. On files with multi-billion-nonzero matrices, that open spiked to **15–24+ GB (nondeterministic)** and OOM-killed the 60 GB Batch container *mid-metadata-read*. Prod symptom: `integrated_all_tissues_object.h5ad` stuck "pending" (job `5cbae667`, exit 137, "container killed due to memory usage").

The investigation showed the validators themselves are **not** the problem.

## Measured peak RSS on the offending file (2.24M × 36.6k; X/raw.X/`ambient_corrected` each ~7.6B nnz = 256 GB if fully resident)

| step | peak RSS | outcome |
|---|---|---|
| CAP validator | 2.71 GB | ✅ completes |
| HCA schema validator (dask-chunked) | 15.42 GB | ✅ completes (26 min) |
| **metadata `read_h5ad(backed="r")`** | **15 → 24+ GB** | ❌ nondeterministic, OOMs |
| **metadata via `read_elem` (this PR)** | **1.37 GB** | ✅ |

## Change

- `read_file_metadata` now reads `obs` + `uns` via targeted `read_elem` and gets shape from the X header — never opening X/raw/layers. Summary + coverage are unchanged (they only need obs/uns; `read_elem(obs)` is anndata's own obs-read path, so output is identical).
- New `matrix_storage` payload field: per-matrix/layer `format, n_obs, n_vars, nnz, data_dtype, index_dtype, on_disk_bytes, resident_bytes`, from a new pure-h5py `hca_validation.h5ad_storage` module (HDF5 header only, no matrix load). Additive + SNS-safe — the tracker's Yup schema ignores unknown keys (persistence tracked in clevercanary/hca-atlas-tracker#1436).
- `h5py` wired as an optional `h5ad` extra on `shared` so the entry-sheet Lambda stays lean.

## Tests

- Swapped the now-bypassed `read_h5ad` mock for a mockable read seam (`_read_metadata_inputs`/`get_matrix_storage`).
- Added `test_read_file_metadata_real_h5ad` exercising the real (unmocked) read path end-to-end.
- New `shared/tests/test_h5ad_storage.py` (incl. the int32-vs-int64 index footprint case).
- Full suite green (28 dataset-validator + 4 shared); pyright clean on changed files.

## Notes

- This **supersedes** the EC2/512 GB direction in clevercanary/hca-atlas-tracker-tf-config#68 — nothing in the pipeline needs >~15 GB once this lands.
- Follow-up #448: reuse the shared helper in `hca-anndata-tools/storage.py` + fix the same backed-read OOM in `get_summary`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: shared/src/hca_validation/h5ad_storage/h5ad_storage.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+from typing import Any, Optional, cast
+
+import h5py
+import numpy as np
+
+_SPARSE_ENCODINGS = ("csr_matrix", "csc_matrix")
+
+
+def _decode(value: Any) -> Any:
+    return value.decode() if isinstance(value, bytes) else value
```
[the remark this answers is not in the document]
Comment (NoopDog) [reply shared/src/hca_validation/h5ad_storage/h5ad_storage.py:27]: Verified this is a non-issue: `numpy.bytes_` is a subclass of Python `bytes`, so `isinstance(value, bytes)` catches it and `_decode(np.bytes_(b'csr_matrix'))` returns `'csr_matrix'`. h5py surfaces fixed-length string attrs as `numpy.bytes_` and variable-length as `str` — both decode correctly, and `get_matrix_storage` already identifies `csr_matrix` on real files. No change needed.

File: services/dataset-validator/src/dataset_validator/main.py [hunk 2]
```diff
@@ -158,11 +163,25 @@ def _json_length_of_report_list(self, report_key: str, message_type: str) -> int
...
+        reads preferentially. Keeping it out keeps the SNS message shape identical
+        to the pre-#447 payload.
+        """
+
+        # The inline SNS message never carries matrix_storage; serialize/truncate a
+        # view of self with it stripped. (No copy in the common case where it's None.)
+        base = self
+        if self.matrix_storage is not None:
+            base = copy.deepcopy(self)
+            base.matrix_storage = None
```
[the remark this answers is not in the document]
Comment (NoopDog) [reply services/dataset-validator/src/dataset_validator/main.py]: Fixed in d7db9d2 — switched to a shallow `copy.copy` (only the matrix_storage field is reassigned; the truncation path deep-copies before mutating tool_reports, so the original is never deep-copied).

File: services/dataset-validator/src/dataset_validator/main.py [hunk 3]
```diff
@@ -471,30 +490,82 @@ def _get_schemaview():
...
+    if "var" in f:
+        var = f["var"]
+        # Defensive: this fallback only runs for files where X lacks a shape
+        # attr (not produced by anndata). Degrade to 0 rather than raising a
+        # cryptic h5py error if the var index is missing/malformed.
+        index_name = var.attrs.get("_index")
+        if isinstance(index_name, bytes):
+            index_name = index_name.decode()
+        if index_name is not None and index_name in var:
+            n_vars = int(var[index_name].shape[0])
```
[the remark this answers is not in the document]
Comment (NoopDog) [reply services/dataset-validator/src/dataset_validator/main.py]: Fixed in d7db9d2 — the var-index node is now guarded with `isinstance(node, h5py.Dataset)`, so the fallback degrades to n_vars=0 instead of raising on a malformed var group. Test extended to cover it.

File: services/dataset-validator/src/dataset_validator/main.py [hunk 4]
```diff
@@ -158,11 +163,28 @@ def _json_length_of_report_list(self, report_key: str, message_type: str) -> int
...
 
     def to_length_limited_json(self, max_length=MAX_SNS_MESSAGE_LENGTH) -> str:
-        """Convert to a JSON string with message lists truncated to ensure that the JSON is below a given length."""
+        """Convert to a JSON string with message lists truncated to ensure that the JSON is below a given length.
+
+        ``matrix_storage`` (#447) is omitted entirely from the inline SNS payload:
+        it can be large (a file with many layers) and is always available in full
+        from the S3 claim-check (write_validation_results_to_s3), which the tracker
+        reads preferentially. Keeping it out keeps the SNS message shape identical
+        to the pre-#447 payload.
```
[the remark this answers is not in the document]
Comment (NoopDog) [reply services/dataset-validator/src/dataset_validator/main.py]: Fixed in 1d40e5c — the SNS path now drops the matrix_storage key entirely via a dedicated `_to_sns_json` serializer (not just sets it to null), so the inline payload shape is truly identical to pre-#447. The full value still goes to the S3 claim-check via `to_json`. Docstring updated to match.

File: shared/src/hca_validation/h5ad_storage/h5ad_storage.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+            "format": encoding,
+            "n_obs": shape[0] if len(shape) == 2 else None,
+            "n_vars": shape[1] if len(shape) == 2 else None,
+            "nnz": nnz,
+            "data_dtype": str(np.dtype(data.dtype)),
+            "index_dtype": str(np.dtype(indices.dtype)),
+            "on_disk_bytes": int(on_disk),
+            "resident_bytes": int(resident),
+        }
+
```
[the remark this answers is not in the document]
Comment (NoopDog) [reply shared/src/hca_validation/h5ad_storage/h5ad_storage.py:81]: Fixed in 1d40e5c — `_matrix_info` now returns None for a sparse node missing data/indices/indptr instead of raising, so one malformed node no longer fails introspection for the whole file. Test added.

File: services/dataset-validator/src/dataset_validator/main.py [hunk 6]
```diff
@@ -213,7 +236,7 @@ def to_length_limited_json(self, max_length=MAX_SNS_MESSAGE_LENGTH) -> str:
                 *threshold_path,
                 [*message_list[:middle_length], _truncated_marker(middle_length, len(message_list))],
             )
-            if len(truncated_message.to_json()) < max_length:
+            if len(truncated_message._to_sns_json()) < max_length:
                 max_valid_length = middle_length
             else:
                 min_invalid_length = middle_length
```
[the remark this answers is not in the document]
Comment (NoopDog) [reply services/dataset-validator/src/dataset_validator/main.py:247]: Good catch — fixed in d79ea6c. The final post-binary-search return now uses `_to_sns_json()` like the rest of the method, so the truncation path also omits matrix_storage and stays under the measured limit. Added a test that forces the binary-search path.

File: shared/src/hca_validation/h5ad_storage/h5ad_storage.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    if isinstance(node, h5py.Dataset):
+        shape = tuple(int(x) for x in node.shape)
+        itemsize = np.dtype(node.dtype).itemsize
+        n_elements = int(np.prod(shape)) if shape else 0
```
[the remark this answers is not in the document]
Comment (NoopDog) [reply shared/src/hca_validation/h5ad_storage/h5ad_storage.py]: Fixed in d79ea6c — dense element count now uses `np.prod(shape, dtype=object)` to keep the product in Python big-ints and avoid fixed-width overflow.

File: shared/pyproject.toml [hunk 8]
```diff
@@ -20,12 +20,20 @@ pydantic = "^2.11.4"
...
+
+[tool.poetry.extras]
+h5ad = ["h5py"]
 
 [tool.poetry.group.dev.dependencies]
 pytest = "^8.3.5"
 pytest-mock = "^3.14.0"
 python-dotenv = "^1.1.0"
 requests = "^2.32.3"
+h5py = "^3.8"
```
[the remark this answers is not in the document]
Comment (NoopDog) [reply shared/pyproject.toml:36]: Verified this doesn't defeat the lean goal: `poetry install --only main` reports h5py as "Skipped … Not required" — a main-only install (what the entry-sheet Lambda gets: shared's main deps, no `h5ad` extra, no dev-group transitivity) does NOT pull h5py. The `optional=false`/`groups=["main","dev"]` marking reflects that h5py is legitimately both an optional *runtime* dep (gated by the `h5ad` extra) and a *test* dep (dev group). Removing the dev dep would force shared's h5ad tests to be skipped on a default `poetry install` unless CI adds `--extras h5ad`, so I'm keeping it. Happy to restructure if preferred.

File: services/dataset-validator/src/dataset_validator/main.py [hunk 9]
```diff
@@ -157,11 +162,26 @@ def _set_report_message_list(self, report_key: str, message_type: str, value: Li
...
+
+        ``matrix_storage`` (#447) can be large (a file with many layers) and is
+        always available in full from the S3 claim-check
+        (write_validation_results_to_s3), which the tracker reads preferentially.
+        Dropping the key keeps the SNS message shape identical to the pre-#447
+        payload.
+        """
+        data = asdict(self)
+        data.pop("matrix_storage", None)
+        return json.dumps(data, separators=(',', ':'))
```
[the remark this answers is not in the document]
Comment (NoopDog) [reply services/dataset-validator/src/dataset_validator/main.py]: Fixed in ac584cb — matrix_storage is now nulled on a shallow copy before asdict, so its payload is never recursively deep-copied just to be dropped.

File: services/dataset-validator/src/dataset_validator/main.py [hunk 10]
```diff
@@ -471,30 +494,84 @@ def _get_schemaview():
...
+        # Defensive: this fallback only runs for files where X lacks a shape
+        # attr (not produced by anndata). Degrade to 0 rather than raising a
+        # cryptic h5py error if the var index is missing/malformed.
+        index_name = var.attrs.get("_index")
+        if isinstance(index_name, bytes):
+            index_name = index_name.decode()
+        if index_name is not None and index_name in var:
+            node = var[index_name]
+            if isinstance(node, h5py.Dataset):
+                n_vars = int(node.shape[0])
```
[the remark this answers is not in the document]
Comment (NoopDog) [reply services/dataset-validator/src/dataset_validator/main.py]: Fixed in ac584cb — added an isinstance(var, h5py.Group) guard so the fallback degrades to n_vars=0 if var is a dataset instead of raising.

File: shared/src/hca_validation/h5ad_storage/h5ad_storage.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+        on_disk = _storage_size(data) + _storage_size(indices) + _storage_size(indptr)
+        return {
+            "format": encoding,
+            "n_obs": shape[0] if len(shape) == 2 else None,
+            "n_vars": shape[1] if len(shape) == 2 else None,
+            "nnz": nnz,
+            "data_dtype": str(np.dtype(data.dtype)),
+            "index_dtype": str(np.dtype(indices.dtype)),
+            "on_disk_bytes": int(on_disk),
+            "resident_bytes": int(resident),
```
[the remark this answers is not in the document]
Comment (NoopDog) [reply shared/src/hca_validation/h5ad_storage/h5ad_storage.py:79]: Fixed in ac584cb — added indptr_dtype alongside index_dtype (they can differ, e.g. int32 indices with int64 indptr), so the storage facts fully describe resident_bytes. Test added for the mixed-dtype case.

File: services/dataset-validator/src/dataset_validator/main.py [hunk 12]
```diff
@@ -471,30 +500,85 @@ def _get_schemaview():
...
+
+def read_file_metadata(
+    file_path: Path,
+) -> Tuple[MetadataSummary, Optional[dict], Optional[dict]]:
+    """Read metadata once and return (summary, coverage, matrix_storage).
+
+    Only obs + uns (+ matrix headers) are read — never the matrices themselves —
+    so this stays bounded (~1 GB) even on files whose X/raw/layers hold billions
+    of nonzeros. Coverage and matrix-storage failures are non-fatal (returned as
+    None); a failure to read obs/uns is fatal (raised through).
```
[the remark this answers is not in the document]
Comment (NoopDog) [reply services/dataset-validator/src/dataset_validator/main.py:520]: Done in b30efad — removed read_metadata() entirely (it was dead in prod and still did the backed read this PR eliminates). Its test now exercises _extract_metadata_summary directly, keeping the edge-case coverage; the unused `import anndata` is dropped too.

File: services/dataset-validator/src/dataset_validator/main.py [hunk 13]
```diff
@@ -471,30 +500,85 @@ def _get_schemaview():
...
-    Coverage failures are non-fatal — they're returned as None so the
-    caller can still publish a summary. Reading the file at all is fatal
-    (raised through), since downstream validators depend on it.
+    X carries a ``shape`` attr when sparse; when dense it's a 2-D dataset.
+    """
+    if "X" in f:
+        x = f["X"]
+        shape = x.attrs.get("shape")
+        if shape is not None and len(shape) == 2:
+            return int(shape[0]), int(shape[1])
```
[the remark this answers is not in the document]
Comment (NoopDog) [reply services/dataset-validator/src/dataset_validator/main.py:474]: Fixed in b30efad — added a non-sequence guard (hasattr __len__) so a malformed scalar shape attr falls through to the var/obs fallback instead of raising on len(). Test added.

File: shared/src/hca_validation/h5ad_storage/h5ad_storage.py [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+    if isinstance(node, h5py.Group) and encoding in _SPARSE_ENCODINGS:
+        # Treat an incomplete sparse node (missing data/indices/indptr) as
+        # unrecognized rather than raising — one malformed node must not fail
+        # introspection for the whole file.
+        if not all(k in node for k in ("data", "indices", "indptr")):
+            return None
+        data = cast(h5py.Dataset, node["data"])
+        indices = cast(h5py.Dataset, node["indices"])
+        indptr = cast(h5py.Dataset, node["indptr"])
+        shape = tuple(int(x) for x in node.attrs.get("shape", ()))
```
[the remark this answers is not in the document]
Comment (NoopDog) [reply shared/src/hca_validation/h5ad_storage/h5ad_storage.py:63]: Fixed in 85b35b7 — rather than guard shape specifically, _matrix_info now wraps parsing in a try/except so any unintrospectable node (bad shape, odd dtype, etc.) degrades to None instead of aborting the whole file. The explicit data/indices/indptr guard stays for clarity; this is the general backstop. Test added for the scalar-shape case.
